File: python/anecho/container.py
"""
anecho model container (``.anecho``) — writer, reader, and integrity check.

Layout (little-endian):
    magic   b"ANEC"            4 bytes
    version u16                container format version (1 or 2)
    hlen    u32                header length in bytes
    header  hlen bytes         UTF-8 JSON
    payload rest of file       tensor data, each tensor 64-byte aligned

Version 2 header::

    {
      "format": "anecho-model/2",
      "model_id": "anecho-openspace-16khz-v1",
      "created": "2026-08-17T...",
      "config": { ...ModelConfig... },
      "graph": { "inputs": [...], "outputs": {...},
                 "nodes": [...], "numerics": {...} },
      "payload_len": <int>,
      "tensors": [ {"name","dtype","shape","offset","nbytes"}, ... ],
      "integrity": { "crc32": <int>, "sha256": "<hex>", "hmac_sha256": "<hex>"|null }
    }

What version 2 adds, and why
----------------------------
**The graph travels with the weights.** Version 1 stored tensors plus a config
dict and left the topology in ``anecho.model`` — the same arrangement as the
vendor's ``.aicmodel``, where the runtime *is* the schema. That is what makes a
foreign model file a reverse-engineering project, and it means an old file
silently misloads against new code. A v2 file names its ops, its wiring and the
numeric conventions that weights cannot express (see ``graph.Numerics``).

**Offsets are payload-relative.** In the vendor format the tensor table holds
absolute file offsets, so nothing may change size — edits are possible only in
place. Here the header is regenerated on every write and offsets are relative to
the payload, so repacking with different shapes is ordinary.

**Integrity is layered.** ``crc32`` is a cheap corruption check that costs
nothing to verify on load; ``sha256`` detects tampering; ``hmac_sha256`` (when
packed with a product secret) detects tampering by anyone without the secret.
The vendor file's two CRCs are corruption checks only — and one of them covers
the other, a detail that silently rejects every naive edit.

**Dtypes are declared.** f32 today, f16/i8 reserved, so quantised models do not
need a new container.

Reading is backward compatible: v1 files still load, with ``graph`` set to None.
"""
from __future__ import annotations

import hashlib
import hmac
import json
import struct
import zlib

import numpy as np

from .errors import ModelInvalidError, ModelVersionUnsupportedError
from .graph import FORMAT, Graph, VadConfig

MAGIC = b"ANEC"
VERSION = 2
SUPPORTED = (1, 2)
ALIGN = 64
HEAD_LEN = 10                       # magic(4) + version(2) + hlen(4)

DTYPES = {"f32": np.float32, "f16": np.float16, "i8": np.int8}

#: Quantisation policies. Per-tensor dtype is what makes mixed precision cheap:
#: the weight matrices carry the size, while BatchNorm statistics and biases —
#: where 8 bits would actually hurt — stay f32.
def _i8_weights(name: str, arr: np.ndarray) -> str:
    return "i8" if (arr.ndim >= 2 and arr.size >= 1024) else "f32"


def _f16_weights(name: str, arr: np.ndarray) -> str:
    return "f16" if (arr.ndim >= 2 and arr.size >= 1024) else "f32"


POLICIES = {"none": None, "i8_weights": _i8_weights, "f16_weights": _f16_weights}


def _canonical(header_without_integrity: dict) -> bytes:
    return json.dumps(header_without_integrity, sort_keys=True,
                      separators=(",", ":")).encode("utf-8")


def _integrity_input(header: dict, payload: bytes) -> bytes:
    h = dict(header)
    h.pop("integrity", None)
    return _canonical(h) + payload


def _pack_payload(tensors: dict, dtype: str, policy=None) -> tuple[bytes, list]:
    payload, entries = bytearray(), []
    for name, arr in tensors.items():
        src = np.ascontiguousarray(arr, dtype=np.float32)
        dt = (policy(name, src) if policy else dtype)
        entry = {"name": name, "dtype": dt, "shape": list(src.shape)}
        if dt == "i8":
            # Symmetric per-tensor quantisation; the scale travels with the tensor
            # so a reader never has to know how it was produced.
            scale = float(np.abs(src).max()) / 127.0 or 1.0
            a = np.clip(np.rint(src / scale), -127, 127).astype(np.int8)
            entry["scale"] = scale
        else:
            a = src.astype(DTYPES[dt])
        payload.extend(b"\0" * (-len(payload) % ALIGN))
        entry["offset"] = len(payload)
        entry["nbytes"] = int(a.nbytes)
        entries.append(entry)
        payload += a.tobytes()
    return bytes(payload), entries


def write_model(path: str, tensors: dict, config: dict, model_id: str, created: str,
                secret: bytes | None = None, graph: Graph | None = None,
                dtype: str = "f32", quantize: str = "none",
                runtime: dict | None = None) -> str:
    """Pack tensors + config (+ graph) into a ``.anecho`` container.

    Passing ``graph`` writes format 2 — a self-describing model. Omitting it
    writes the legacy format 1, where the loader must already know the
    architecture; new models should always carry a graph.
    """
    if dtype not in DTYPES:
        raise ValueError(f"unsupported dtype {dtype!r}; known: {sorted(DTYPES)}")
    if quantize not in POLICIES:
        raise ValueError(f"unknown quantisation policy {quantize!r}; "
                         f"known: {sorted(POLICIES)}")
    payload, entries = _pack_payload(tensors, dtype, POLICIES[quantize])

    if graph is not None:
        graph.validate(tensor_names=set(tensors))
        dead = graph.unused_tensors(set(tensors))
        if dead:
            raise ValueError(f"{len(dead)} tensor(s) no node consumes: {dead[:5]} — "
                             "either wire them up or leave them out")

    header = {
        "format": FORMAT if graph is not None else "anecho-model/1",
        "model_id": model_id,
        "created": created,
        "config": config,
        "payload_len": len(payload),
        "tensors": entries,
    }
    if graph is not None:
        header["graph"] = graph.to_json()
    if runtime:
        if "vad" in runtime:                     # validated, not free-form JSON
            runtime = dict(runtime)
            runtime["vad"] = VadConfig.from_json(runtime["vad"]).to_json()
        header["runtime"] = runtime

    digest = hashlib.sha256(_integrity_input(header, payload)).hexdigest()
    mac = hmac.new(secret, bytes.fromhex(digest), hashlib.sha256).hexdigest() if secret else None
    integrity = {"sha256": digest, "hmac_sha256": mac}
    if graph is not None:
        integrity["crc32"] = zlib.crc32(payload) & 0xFFFFFFFF
    header["integrity"] = integrity

    hb = _canonical(header)
    with open(path, "wb") as f:
        f.write(MAGIC)
        f.write(struct.pack("<H", VERSION if graph is not None else 1))
        f.write(struct.pack("<I", len(hb)))
        f.write(hb)
        f.write(payload)
    return path


def read_model(path: str, secret: bytes | None = None, require_hmac: bool = False,
               require_graph: bool = False):
    """Parse and verify a ``.anecho`` container.

    Returns ``(header, tensors, graph)`` — ``graph`` is None for v1 files.
    Raises ``ModelInvalidError`` on any integrity failure.
    """
    with open(path, "rb") as f:
        d = f.read()
    if len(d) < HEAD_LEN or d[:4] != MAGIC:
        raise ModelInvalidError("bad magic — not an .anecho container")
    ver = struct.unpack_from("<H", d, 4)[0]
    if ver not in SUPPORTED:
        raise ModelVersionUnsupportedError(
            f"container version {ver}; this build reads {SUPPORTED}")
    hlen = struct.unpack_from("<I", d, 6)[0]
    if HEAD_LEN + hlen > len(d):
        raise ModelInvalidError("header length runs past the end of the file")
    try:
        header = json.loads(d[HEAD_LEN:HEAD_LEN + hlen].decode("utf-8"))
    except Exception as e:
        raise ModelInvalidError(f"header parse: {e}")

    payload = d[HEAD_LEN + hlen:]
    if len(payload) != header.get("payload_len", -1):
        raise ModelInvalidError("payload length mismatch")

    integ = header.get("integrity") or {}
    crc = integ.get("crc32")                     # cheap check first
    if crc is not None and (zlib.crc32(payload) & 0xFFFFFFFF) != crc:
        raise ModelInvalidError("payload crc32 mismatch (file is corrupted)")
    digest = hashlib.sha256(_integrity_input(header, payload)).hexdigest()
    if not hmac.compare_digest(digest, integ.get("sha256", "")):
        raise ModelInvalidError("sha256 integrity mismatch (file was modified)")
    stored_mac = integ.get("hmac_sha256")
    if secret is not None and stored_mac:
        expect = hmac.new(secret, bytes.fromhex(digest), hashlib.sha256).hexdigest()
        if not hmac.compare_digest(expect, stored_mac):
            raise ModelInvalidError("hmac integrity mismatch (wrong secret or tampered)")
    if require_hmac and not (secret is not None and stored_mac):
        raise ModelInvalidError("hmac required but not present/verified")

    tensors = {}
    for t in header["tensors"]:
        dt = DTYPES.get(t.get("dtype", "f32"))
        if dt is None:
            raise ModelInvalidError(f"tensor {t['name']}: unknown dtype {t.get('dtype')!r}")
        o, n = t["offset"], t["nbytes"]
        if o + n > len(payload):
            raise ModelInvalidError(f"tensor {t['name']} runs past the payload")
        arr = np.frombuffer(payload[o:o + n], dtype=dt).reshape(t["shape"])
        arr = np.array(arr)                      # copy out of the read-only buffer
        if "scale" in t:                         # quantised: dequantise on load
            arr = arr.astype(np.float32) * np.float32(t["scale"])
        tensors[t["name"]] = arr

    if "vad" in (header.get("runtime") or {}):   # fail on load, not at first use
        VadConfig.from_json(header["runtime"]["vad"])

    graph = None
    if "graph" in header:
        graph = Graph.from_json(header["graph"]).validate(tensor_names=set(tensors))
    elif require_graph:
        raise ModelInvalidError(
            f"{path} is format {header.get('format')} and carries no graph; "
            "repack it with a graph before loading in a graph-only runtime")
    return header, tensors, graph


def inspect(path: str) -> str:
    """Human-readable summary — a model you can read without running it."""
    header, tensors, graph = read_model(path)
    n_params = sum(int(np.prod(t["shape"])) for t in header["tensors"])
    lines = [f"{header['model_id']}  [{header['format']}]",
             f"  created  {header.get('created')}",
             f"  tensors  {len(tensors)}  ({n_params:,} parameters, "
             f"{header['payload_len']:,} bytes)",
             f"  integrity crc32={'yes' if (header.get('integrity') or {}).get('crc32') else 'no'} "
             f"sha256=yes hmac={'yes' if (header.get('integrity') or {}).get('hmac_sha256') else 'no'}"]
    if graph is not None:
        lines.append(f"  numerics {json.dumps(graph.numerics.to_json(), sort_keys=True)}")
        lines.append("  graph:")
        lines += ["    " + ln for ln in graph.describe().splitlines()]
    else:
        lines.append("  graph:   ABSENT (format 1 — the loader must know the architecture)")
    return "\n".join(lines)

File: node/src/container.js
'use strict';
// Reader for the .anecho container. The format is deliberately simple enough that
// a second implementation is a day's work rather than a reverse-engineering
// project — which is the whole point of shipping the graph inside the file.
const fs = require('fs');
const crypto = require('crypto');
const zlib = require('zlib');

const MAGIC = Buffer.from('ANEC', 'ascii');
const HEAD_LEN = 10;
const SUPPORTED = [1, 2];
const DTYPES = { f32: Float32Array, f16: Uint16Array, i8: Int8Array };

/** IEEE half -> float. The format declares f16, so a loader that cannot read it
 *  is not a loader for the format; JS has no Float16Array before Node 24. */
function half(bits) {
  const s = bits >> 15 ? -1 : 1, e = (bits >> 10) & 0x1f, m = bits & 0x3ff;
  if (e === 0) return s * m * 2 ** -24;                       // subnormal, and zero
  if (e === 31) return m ? NaN : s * Infinity;
  return s * (1 + m / 1024) * 2 ** (e - 15);
}

/** Remove the top-level "integrity" member from the *stored* header bytes.
 *
 * The digest covers `canonical(header without integrity)`, and re-serialising the
 * parsed header to get that back is a trap: Python writes floats as `1e-05` and
 * `40.0` where JSON.stringify gives `1e-5` and `40`, so the bytes differ and every
 * file looks tampered with. The stored header already *is* the canonical form with
 * sorted keys, so deleting one member's text — key, value, and the adjacent comma —
 * yields exactly what the writer hashed, with no float formatting involved.
 */
function stripIntegrity(text) {
  const key = '"integrity":';
  let depth = 0, at = -1;
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    if (depth === 1 && c === '"' && text.startsWith(key, i)) { at = i; break; }
    if (c === '"') {                                  // skip strings wholesale
      i++;
      while (i < text.length && (text[i] !== '"' || text[i - 1] === '\\')) i++;
      continue;
    }
    if (c === '{' || c === '[') depth++;
    else if (c === '}' || c === ']') depth--;
  }
  if (at < 0) throw new Error('header carries no integrity member');
  let j = at + key.length, d = 0;
  do {
    const c = text[j];
    if (c === '"') { j++; while (j < text.length && (text[j] !== '"' || text[j - 1] === '\\')) j++; }
    else if (c === '{' || c === '[') d++;
    else if (c === '}' || c === ']') d--;
    j++;
  } while (d > 0 && j < text.length);
  let start = at, end = j;
  if (text[start - 1] === ',') start -= 1;
  else if (text[end] === ',') end += 1;
  return text.slice(0, start) + text.slice(end);
}


function readModel(path, { secret = null, requireGraph = false } = {}) {
  const buf = fs.readFileSync(path);
  if (buf.length < HEAD_LEN || !buf.subarray(0, 4).equals(MAGIC)) {
    throw new Error('bad magic — not an .anecho container');
  }
  const version = buf.readUInt16LE(4);
  if (!SUPPORTED.includes(version)) {
    throw new Error(`container version ${version}; this build reads ${SUPPORTED}`);
  }
  const hlen = buf.readUInt32LE(6);
  if (HEAD_LEN + hlen > buf.length) throw new Error('header runs past the end of the file');
  const header = JSON.parse(buf.subarray(HEAD_LEN, HEAD_LEN + hlen).toString('utf8'));
  const payload = buf.subarray(HEAD_LEN + hlen);
  if (payload.length !== header.payload_len) throw new Error('payload length mismatch');

  const integ = header.integrity || {};
  if (integ.crc32 !== undefined && (zlib.crc32 ? zlib.crc32(payload) : crc32(payload)) !== integ.crc32) {
    throw new Error('payload crc32 mismatch (file is corrupted)');
  }
  const headerText = buf.subarray(HEAD_LEN, HEAD_LEN + hlen).toString('utf8');
  const digest = crypto.createHash('sha256')
    .update(Buffer.from(stripIntegrity(headerText), 'utf8')).update(payload).digest('hex');
  if (digest !== integ.sha256) throw new Error('sha256 integrity mismatch (file was modified)');
  if (secret && integ.hmac_sha256) {
    const expect = crypto.createHmac('sha256', secret)
      .update(Buffer.from(digest, 'hex')).digest('hex');
    if (expect !== integ.hmac_sha256) throw new Error('hmac mismatch (wrong secret or tampered)');
  }

  // Tensor offsets are 64-byte aligned inside the payload, but the payload itself
  // starts at an arbitrary file offset, so a typed-array view over the file buffer
  // would fault. One copy at load time buys aligned views for every tensor.
  const store = new ArrayBuffer(payload.length);
  new Uint8Array(store).set(payload);

  const tensors = new Map();
  for (const t of header.tensors) {
    const Ctor = DTYPES[t.dtype];
    if (!Ctor) throw new Error(`tensor ${t.name}: unsupported dtype ${t.dtype}`);
    const view = new Ctor(store, t.offset, t.nbytes / Ctor.BYTES_PER_ELEMENT);
    let data = t.dtype === 'f16' ? Float32Array.from(view, half) : Float32Array.from(view);
    if (t.scale !== undefined) for (let i = 0; i < data.length; i++) data[i] *= t.scale;
    tensors.set(t.name, { d: data, s: t.shape.slice() });
  }

  const graph = header.graph || null;
  if (!graph && requireGraph) {
    throw new Error(`${path} is format ${header.format} and carries no graph`);
  }
  if (graph) validateGraph(graph, tensors);
  return { header, tensors, graph };
}

function validateGraph(graph, tensors) {
  const available = new Set(graph.inputs);
  for (const node of graph.nodes) {
    for (const v of node.inputs) {
      if (!available.has(v)) {
        throw new Error(`${node.name}: input '${v}' is not produced by any earlier node`);
      }
    }
    for (const v of node.outputs) available.add(v);
    for (const [role, name] of Object.entries(node.params || {})) {
      if (!tensors.has(name)) throw new Error(`${node.name}: parameter ${role}='${name}' missing`);
    }
  }
  for (const [pub, v] of Object.entries(graph.outputs)) {
    if (!available.has(v)) throw new Error(`output '${pub}' refers to '${v}', which nothing produces`);
  }
}

// zlib.crc32 exists from Node 20.15; keep a fallback so older runtimes still load.
let TABLE = null;
function crc32(buf) {
  if (!TABLE) {
    TABLE = new Int32Array(256);
    for (let i = 0; i < 256; i++) {
      let c = i;
      for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
      TABLE[i] = c;
    }
  }
  let c = 0 ^ -1;
  for (let i = 0; i < buf.length; i++) c = (c >>> 8) ^ TABLE[(c ^ buf[i]) & 0xff];
  return (c ^ -1) >>> 0;
}

module.exports = { readModel, stripIntegrity };
